File: client/src/main/java/com/securechat/client/SecureConnection.java
package com.securechat.client;

import com.google.gson.Gson;
import com.securechat.common.crypto.AESUtil;
import com.securechat.common.crypto.RSAUtil;
import com.securechat.common.dto.MessageDTO;
import com.securechat.common.dto.UserDTO;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.crypto.SecretKey;
import javax.net.ssl.*;
import java.io.*;
import java.security.KeyPair;
import java.security.KeyStore;
import java.util.Base64;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.function.Consumer;

/**
 * SecureConnection – Quản lý kết nối SSL/TLS tới server và mã hoá AES.
 *
 * Luồng kết nối:
 * 1. Load truststore (CA cert) → Verify server certificate
 * 2. Thiết lập SSLSocket → server:8443
 * 3. Gửi thông tin đăng nhập
 * 4. Nhận RSA Public Key của server
 * 5. Sinh AES Key 256-bit → mã hoá bằng RSA → gửi lên server
 * 6. Bắt đầu vòng lặp nhận/gửi tin nhắn mã hoá AES
 */
public class SecureConnection {

    private static final Logger log = LoggerFactory.getLogger(SecureConnection.class);

    private static final String TRUSTSTORE_PATH = "/client.truststore";
    private static final String TRUSTSTORE_PASS = "trust_password_2024";

    private final String  serverHost;
    private final int     serverPort;
    private final Gson    gson = new Gson();

    // Trạng thái kết nối
    private SSLSocket     socket;
    private PrintWriter   out;
    private BufferedReader in;

    private SecretKey aesKey;       // Khoá AES phiên chat
    private KeyPair   clientKeyPair; // RSA keypair của client
    private UserDTO   loggedInUser;

    // Callbacks
    private Consumer<MessageDTO> onMessageReceived;
    private Consumer<String>     onConnectionError;

    // Queue để gửi tin nhắn (thread-safe)
    private final BlockingQueue<MessageDTO> sendQueue = new LinkedBlockingQueue<>();

    private volatile boolean connected = false;

    // ────────────────────────────────────────────────────────────

    public SecureConnection(String serverHost, int serverPort) {
        this.serverHost = serverHost;
        this.serverPort = serverPort;
    }

    // ────────────────────────────────────────────────────────────
    // Kết nối và đăng nhập
    // ────────────────────────────────────────────────────────────

    /**
     * Kết nối tới server và thực hiện login + key exchange.
     * @return UserDTO nếu thành công, null nếu thất bại
     */
    public UserDTO connect(String username, String password) throws Exception {
        // Bước 1: Tạo SSLSocket với truststore
        SSLSocket sslSocket = createSSLSocket();
        this.socket = sslSocket;
        sslSocket.startHandshake();
        log.info("SSL Handshake thành công với {}", sslSocket.getInetAddress());

        // Bước 2: Khởi tạo streams
        this.out = new PrintWriter(new BufferedWriter(
                new OutputStreamWriter(sslSocket.getOutputStream())), true);
        this.in  = new BufferedReader(
                new InputStreamReader(sslSocket.getInputStream()));

        // Bước 3: Sinh RSA Keypair cho client
        this.clientKeyPair = RSAUtil.generateKeyPair();
        log.info("Đã sinh RSA Keypair cho client");

        // Bước 4: Gửi thông tin đăng nhập
        MessageDTO loginMsg = MessageDTO.loginRequest(username, password);
        sendRaw(loginMsg);

        // Bước 5: Nhận phản hồi đăng nhập + RSA Public Key của server
        String responseJson = in.readLine();
        MessageDTO loginResponse = gson.fromJson(responseJson, MessageDTO.class);

        if (loginResponse.getType() != MessageDTO.Type.LOGIN_RESPONSE) {
            throw new IOException("Server không phản hồi LOGIN_RESPONSE");
        }

        String content = loginResponse.getPlainContent();
        if (content.startsWith("FAIL:")) {
            log.warn("Đăng nhập thất bại: {}", content.substring(5));
            disconnect();
            return null;
        }

        // Lấy server RSA Public Key
        String serverPublicKeyBase64 = loginResponse.getEncryptedContent();
        java.security.PublicKey serverPublicKey = RSAUtil.base64ToPublicKey(serverPublicKeyBase64);

        // Điền thông tin user
        this.loggedInUser = new UserDTO();
        loggedInUser.setId(loginResponse.getSenderId());
        loggedInUser.setUsername(username);

        log.info("Đăng nhập thành công! UserId={}", loggedInUser.getId());

        // Bước 6: Gửi RSA Public Key của client lên server
        MessageDTO clientKeyMsg = new MessageDTO(MessageDTO.Type.KEY_EXCHANGE);
        clientKeyMsg.setPlainContent(RSAUtil.publicKeyToBase64(clientKeyPair.getPublic()));
        sendRaw(clientKeyMsg);

        // Bước 7: Sinh AES Key và mã hoá bằng RSA Public Key của server
        this.aesKey = AESUtil.generateKey();
        byte[] encryptedAesKey = RSAUtil.encrypt(aesKey.getEncoded(), serverPublicKey);

        MessageDTO aesKeyMsg = new MessageDTO(MessageDTO.Type.KEY_EXCHANGE);
        aesKeyMsg.setPlainContent(Base64.getEncoder().encodeToString(encryptedAesKey));
        sendRaw(aesKeyMsg);

        log.info("Key exchange hoàn thành! AES-256-GCM session key đã thiết lập.");

        // Bước 8: Khởi động vòng lặp nhận tin nhắn
        this.connected = true;
        startReceiveLoop();
        startSendLoop();

        return loggedInUser;
    }

    // ────────────────────────────────────────────────────────────
    // Gửi tin nhắn
    // ────────────────────────────────────────────────────────────

    /**
     * Gửi tin nhắn văn bản (sẽ được mã hoá AES tự động).
     */
    public void sendText(int roomId, String plaintext) throws Exception {
        if (!connected || aesKey == null) throw new IllegalStateException("Chưa kết nối!");

        String encrypted = AESUtil.encrypt(plaintext, aesKey);
        MessageDTO msg = MessageDTO.textMessage(
                loggedInUser.getId(), loggedInUser.getUsername(), roomId, encrypted);

        sendQueue.offer(msg);
    }

    /**
     * Gửi file (bytes đã đọc từ disk).
     */
    public void sendFile(int roomId, String fileName, byte[] fileBytes) throws Exception {
        if (!connected || aesKey == null) throw new IllegalStateException("Chưa kết nối!");

        byte[] encryptedBytes = AESUtil.encryptBytes(fileBytes, aesKey);
        String base64Encrypted = Base64.getEncoder().encodeToString(encryptedBytes);

        MessageDTO msg = MessageDTO.fileMessage(
                loggedInUser.getId(), loggedInUser.getUsername(),
                roomId, base64Encrypted, fileName, fileBytes.length);

        sendQueue.offer(msg);
    }

    /**
     * Giải mã tin nhắn nhận được (dùng AES key của phiên).
     */
    public String decryptMessage(String encryptedContent) {
        try {
            return AESUtil.decrypt(encryptedContent, aesKey);
        } catch (Exception e) {
            log.error("Lỗi giải mã tin nhắn: {}", e.getMessage());
            return "[Không thể giải mã]";
        }
    }

    /**
     * Giải mã file nhận được.
     */
    public byte[] decryptFile(String base64Encrypted) {
        try {
            byte[] encryptedBytes = Base64.getDecoder().decode(base64Encrypted);
            return AESUtil.decryptBytes(encryptedBytes, aesKey);
        } catch (Exception e) {
            log.error("Lỗi giải mã file: {}", e.getMessage());
            return null;
        }
    }

    // ────────────────────────────────────────────────────────────
    // Vòng lặp nhận / gửi
    // ────────────────────────────────────────────────────────────

    private void startReceiveLoop() {
        Thread receiveThread = new Thread(() -> {
            try {
                String line;
                while (connected && (line = in.readLine()) != null) {
                    try {
                        MessageDTO msg = gson.fromJson(line, MessageDTO.class);
                        if (onMessageReceived != null) {
                            onMessageReceived.accept(msg);
                        }
                    } catch (Exception e) {
                        log.error("Lỗi parse tin nhắn: {}", e.getMessage());
                    }
                }
            } catch (IOException e) {
                if (connected) {
                    log.error("Mất kết nối server: {}", e.getMessage());
                    if (onConnectionError != null) {
                        onConnectionError.accept("Mất kết nối server: " + e.getMessage());
                    }
                }
            } finally {
                connected = false;
            }
        }, "ReceiveThread");
        receiveThread.setDaemon(true);
        receiveThread.start();
    }

    private void startSendLoop() {
        Thread sendThread = new Thread(() -> {
            while (connected) {
                try {
                    MessageDTO msg = sendQueue.take();
                    sendRaw(msg);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
        }, "SendThread");
        sendThread.setDaemon(true);
        sendThread.start();
    }

    private synchronized void sendRaw(MessageDTO msg) {
        if (out != null) {
            out.println(gson.toJson(msg));
        }
    }

    // ────────────────────────────────────────────────────────────
    // Tạo SSLSocket với truststore
    // ────────────────────────────────────────────────────────────

    private SSLSocket createSSLSocket() throws Exception {
        KeyStore trustStore = KeyStore.getInstance("JKS");
        try (InputStream is = SecureConnection.class.getResourceAsStream(TRUSTSTORE_PATH)) {
            if (is == null) {
                throw new RuntimeException(
                    "Không tìm thấy client.truststore! Hãy chạy certs/generate_certs.bat trước.");
            }
            trustStore.load(is, TRUSTSTORE_PASS.toCharArray());
        }

        TrustManagerFactory tmf = TrustManagerFactory.getInstance(
                TrustManagerFactory.getDefaultAlgorithm());
        tmf.init(trustStore);

        SSLContext sslContext = SSLContext.getInstance("TLS");
        sslContext.init(null, tmf.getTrustManagers(), null);

        SSLSocketFactory ssf = sslContext.getSocketFactory();
        SSLSocket socket = (SSLSocket) ssf.createSocket(serverHost, serverPort);
        socket.setEnabledProtocols(new String[]{"TLSv1.3", "TLSv1.2"});

        return socket;
    }

    // ────────────────────────────────────────────────────────────
    // Callbacks & Utils
    // ────────────────────────────────────────────────────────────

    public void setOnMessageReceived(Consumer<MessageDTO> callback) {
        this.onMessageReceived = callback;
    }

    public void setOnConnectionError(Consumer<String> callback) {
        this.onConnectionError = callback;
    }

    public void disconnect() {
        connected = false;
        try {
            if (socket != null && !socket.isClosed()) socket.close();
        } catch (IOException e) {
            log.error("Lỗi đóng kết nối: {}", e.getMessage());
        }
    }

    public boolean isConnected()     { return connected; }
    public UserDTO getLoggedInUser() { return loggedInUser; }
    public SecretKey getAesKey()     { return aesKey; }
}

File: SecureChat.Client/SecureConnection.cs
using System;
using System.IO;
using System.Net.Security;
using System.Net.Sockets;
using System.Security.Authentication;
using System.Security.Cryptography.X509Certificates;
using System.Text;
using System.Text.Json;
using System.Threading;
using SecureChat.Common.DTO;
using SecureChat.Common.Crypto;

namespace SecureChat.Client
{
    public class SecureConnection
    {
        private readonly string _serverHost;
        private readonly int _serverPort;

        private TcpClient? _socket;
        private SslStream? _sslStream;
        private StreamReader? _reader;
        private StreamWriter? _writer;

        private byte[]? _aesKey; // Khóa phiên AES-256
        private RSAUtil.RsaKeyPair? _clientKeyPair; // Cặp khóa RSA của client
        private UserDTO? _loggedInUser;
        private readonly object _sendLock = new();
        private Thread? _receiveThread;

        private volatile bool _connected = false;

        public bool IsConnected => _connected;
        public UserDTO? LoggedInUser => _loggedInUser;
        public byte[]? AesKey => _aesKey;

        // Sự kiện callback nhận tin nhắn
        public event Action<MessageDTO>? MessageReceived;
        public event Action<string>? ConnectionError;

        public SecureConnection(string serverHost, int serverPort)
        {
            _serverHost = serverHost;
            _serverPort = serverPort;
        }

        /// <summary>
        /// Kết nối đến Server và thực hiện Handshake (Đăng nhập -> Trao đổi khóa).
        /// </summary>
        public UserDTO? Connect(string username, string password)
        {
            try
            {
                _socket = new TcpClient();
                _socket.Connect(_serverHost, _serverPort);

                // Khởi tạo SslStream với bộ lọc xác thực chứng chỉ CA tùy chỉnh
                _sslStream = new SslStream(
                    _socket.GetStream(),
                    false,
                    new RemoteCertificateValidationCallback(ValidateServerCertificate),
                    null
                );

                // Thực hiện SSL Handshake
                _sslStream.AuthenticateAsClient(_serverHost);

                _reader = new StreamReader(_sslStream, Encoding.UTF8);
                _writer = new StreamWriter(_sslStream, Encoding.UTF8) { AutoFlush = true };

                // 1. Sinh cặp khóa RSA của Client
                _clientKeyPair = RSAUtil.GenerateKeyPair();

                // 2. Gửi thông tin đăng nhập dạng plaintext DTO qua TLS
                var loginMsg = MessageDTO.LoginRequest(username, password);
                SendRaw(loginMsg);

                // 3. Nhận phản hồi đăng nhập chứa RSA Public Key của Server
                string? responseJson = _reader.ReadLine();
                if (string.IsNullOrEmpty(responseJson)) throw new IOException("Server đóng kết nối đột ngột.");

                var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
                var loginResponse = JsonSerializer.Deserialize<MessageDTO>(responseJson, options);
                if (loginResponse == null || loginResponse.Type != MessageDTO.MessageType.LOGIN_RESPONSE)
                {
                    throw new IOException("Phản hồi từ server không đúng giao thức đăng nhập.");
                }

                string content = loginResponse.PlainContent;
                if (content.StartsWith("FAIL:"))
                {
                    Disconnect();
                    throw new Exception(content.Substring(5));
                }

                string serverPublicKey = loginResponse.EncryptedContent;
                
                _loggedInUser = new UserDTO
                {
                    Id = loginResponse.SenderId,
                    Username = username,
                    DisplayName = loginResponse.SenderUsername,
                    AvatarBase64 = loginResponse.FileName ?? string.Empty,
                    Role = string.IsNullOrWhiteSpace(loginResponse.UserRole) ? "USER" : loginResponse.UserRole,
                    IsActive = loginResponse.UserIsActive
                };

                // 4. Gửi RSA Public Key của Client lên Server
                var clientKeyMsg = new MessageDTO(MessageDTO.MessageType.KEY_EXCHANGE)
                {
                    PlainContent = _clientKeyPair.PublicKey
                };
                SendRaw(clientKeyMsg);

                // 5. Sinh khóa AES-256 ngẫu nhiên
                _aesKey = AESUtil.GenerateKey();

                // Mã hóa khóa AES bằng RSA Public Key của Server và gửi đi
                byte[] encryptedAesKey = RSAUtil.Encrypt(_aesKey, serverPublicKey);
                var aesKeyMsg = new MessageDTO(MessageDTO.MessageType.KEY_EXCHANGE)
                {
                    PlainContent = Convert.ToBase64String(encryptedAesKey)
                };
                SendRaw(aesKeyMsg);

                _connected = true;

                return _loggedInUser;
            }
            catch (Exception ex)
            {
                Disconnect();
                throw new Exception($"Lỗi kết nối: {ex.Message}");
            }
        }

        /// <summary>
        /// Gửi tin nhắn văn bản (sẽ được tự động mã hóa AES-256-GCM).
        /// </summary>
        public void SendText(int roomId, string plaintext)
        {
            if (!_connected || _aesKey == null) throw new InvalidOperationException("Chưa kết nối tới server.");

            string encrypted = AESUtil.Encrypt(plaintext, _aesKey);
            var msg = MessageDTO.TextMessage(_loggedInUser!.Id, _loggedInUser.Username, roomId, encrypted);
            SendRaw(msg);
        }

        /// <summary>
        /// Gửi file (bytes) bằng cách mở kết nối phụ cổng 8444, mã hóa và truyền.
        /// </summary>
        public void SendFile(int roomId, string fileName, byte[] fileBytes)
        {
            if (!_connected || _aesKey == null) throw new InvalidOperationException("Chưa kết nối tới server.");

            // 1. Mã hóa file bytes bằng khóa phiên AES
            byte[] encryptedFileBytes = AESUtil.EncryptBytes(fileBytes, _aesKey);

            // 2. Kết nối tới cổng truyền file của server (8444)
            using (var fileSocket = new TcpClient())
            {
                fileSocket.Connect(_serverHost, 8444);

                using (var fileSslStream = new SslStream(
                    fileSocket.GetStream(),
                    false,
                    new RemoteCertificateValidationCallback(ValidateServerCertificate),
                    null))
                {
                    fileSslStream.AuthenticateAsClient(_serverHost);

                    using (var writer = new BinaryWriter(fileSslStream, Encoding.UTF8))
                    {
                        // Gửi metadata trước
                        writer.Write(roomId);
                        writer.Write(_loggedInUser!.Id);
                        writer.Write(fileName);
                        writer.Write(fileBytes.LongLength);
                        writer.Write(encryptedFileBytes.LongLength);

                        // Gửi toàn bộ dữ liệu file đã mã hóa
                        writer.Write(encryptedFileBytes);
                        writer.Flush();

                        // Chờ phản hồi OK từ server
                        using (var reader = new BinaryReader(fileSslStream, Encoding.UTF8))
                        {
                            string response = reader.ReadString();
                            if (response != "OK")
                            {
                                throw new Exception("Server từ chối nhận file.");
                            }
                        }
                    }
                }
            }
        }

        /// <summary>
        /// Giải mã tin nhắn văn bản nhận được bằng khóa phiên AES.
        /// </summary>
        public string DecryptMessage(string encryptedContent)
        {
            if (_aesKey == null) return "[Chưa cấu hình khóa giải mã]";
            try
            {
                return AESUtil.Decrypt(encryptedContent, _aesKey);
            }
            catch (Exception)
            {
                return "[Không thể giải mã tin nhắn]";
            }
        }

        /// <summary>
        /// Giải mã mảng bytes file nhận được bằng khóa phiên AES.
        /// </summary>
        public byte[]? DecryptFile(string base64Encrypted)
        {
            if (_aesKey == null) return null;
            try
            {
                byte[] encryptedBytes = Convert.FromBase64String(base64Encrypted);
                return AESUtil.DecryptBytes(encryptedBytes, _aesKey);
            }
            catch (Exception)
            {
                return null;
            }
        }

        public void StartReceiveLoop()
        {
            if (_receiveThread != null && _receiveThread.IsAlive)
            {
                return;
            }

            _receiveThread = new Thread(() =>
            {
                var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
                try
                {
                    string? line;
                    while (_connected && _reader != null && (line = _reader.ReadLine()) != null)
                    {
                        try
                        {
                            var msg = JsonSerializer.Deserialize<MessageDTO>(line, options);
                            if (msg != null)
                            {
                                MessageReceived?.Invoke(msg);
                            }
                        }
                        catch (Exception ex)
                        {
                            Console.WriteLine($"[SecureConnection] Lỗi parse tin nhắn: {ex.Message}");
                        }
                    }
                }
                catch (Exception ex)
                {
                    if (_connected)
                    {
                        ConnectionError?.Invoke($"Mất kết nối server: {ex.Message}");
                    }
                }
                finally
                {
                    Disconnect();
                }
            })
            {
                IsBackground = true,
                Name = "ServerReceive"
            };
            _receiveThread.Start();
        }

        public void SendMessage(MessageDTO msg)
        {
            lock (_sendLock)
            {
                if (_writer != null)
                {
                    string json = JsonSerializer.Serialize(msg);
                    _writer.WriteLine(json);
                }
            }
        }

        public void RequestAdminUserList()
        {
            SendMessage(new MessageDTO(MessageDTO.MessageType.ADMIN_LIST_USERS));
        }

        public void CreateUser(string username, string password, string displayName)
        {
            string payload = JsonSerializer.Serialize(new
            {
                Username = username,
                Password = password,
                DisplayName = displayName
            });

            SendMessage(new MessageDTO(MessageDTO.MessageType.ADMIN_CREATE_USER)
            {
                PlainContent = payload
            });
        }

        public void UpdateUser(int userId, string username, string password, string displayName)
        {
            string payload = JsonSerializer.Serialize(new
            {
                UserId = userId,
                Username = username,
                Password = password,
                DisplayName = displayName
            });

            SendMessage(new MessageDTO(MessageDTO.MessageType.ADMIN_UPDATE_USER)
            {
                TargetUserId = userId,
                PlainContent = payload
            });
        }

        public void DeleteUser(int userId)
        {
            SendMessage(new MessageDTO(MessageDTO.MessageType.ADMIN_DELETE_USER)
            {
                TargetUserId = userId
            });
        }

        public void SetUserActive(int userId, bool isActive)
        {
            SendMessage(new MessageDTO(MessageDTO.MessageType.ADMIN_SET_ACTIVE)
            {
                TargetUserId = userId,
                PlainContent = isActive.ToString()
            });
        }

        private void SendRaw(MessageDTO msg)
        {
            lock (_sendLock)
            {
                if (_writer != null)
                {
                    string json = JsonSerializer.Serialize(msg);
                    _writer.WriteLine(json);
                }
            }
        }

        private static bool ValidateServerCertificate(
            object sender,
            X509Certificate? certificate,
            X509Chain? chain,
            SslPolicyErrors sslPolicyErrors)
        {
            if (sslPolicyErrors == SslPolicyErrors.None)
                return true;

            string certsDir = GetCertsDirectory();
            string caPath = Path.Combine(certsDir, "ca.crt");
            if (!File.Exists(caPath))
            {
                // Nếu chưa có file ca.crt (ví dụ lúc khởi động đồng thời), chấp nhận tạm
                return false;
            }

            try
            {
                var caCert = new X509Certificate2(caPath);

                var chain2 = new X509Chain();
                chain2.ChainPolicy.ExtraStore.Add(caCert);
                chain2.ChainPolicy.VerificationFlags = X509VerificationFlags.AllowUnknownCertificateAuthority;
                chain2.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;

                if (certificate != null)
                {
                    var serverCert2 = new X509Certificate2(certificate);
                    bool isValid = chain2.Build(serverCert2);

                    if (!isValid)
                    {
                        foreach (var element in chain2.ChainElements)
                        {
                            if (element.Certificate.Thumbprint == caCert.Thumbprint)
                            {
                                return true;
                            }
                        }
                        return false;
                    }

                    var chainRoot = chain2.ChainElements[chain2.ChainElements.Count - 1].Certificate;
                    return chainRoot.Thumbprint == caCert.Thumbprint;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[SecureConnection] Lỗi xác thực CA: {ex.Message}");
            }

            return false;
        }

        private static string GetCertsDirectory()
        {
            string baseDir = AppDomain.CurrentDomain.BaseDirectory;
            DirectoryInfo? dir = new DirectoryInfo(baseDir);
            while (dir != null)
            {
                if (Directory.Exists(Path.Combine(dir.FullName, "SecureChat.Client")) && 
                    Directory.Exists(Path.Combine(dir.FullName, "SecureChat.Server")))
                {
                    string certsDir = Path.Combine(dir.FullName, "certs");
                    if (!Directory.Exists(certsDir))
                    {
                        Directory.CreateDirectory(certsDir);
                    }
                    return certsDir;
                }
                dir = dir.Parent;
            }
            return Path.Combine(baseDir, "certs");
        }

        public void Disconnect()
        {
            _connected = false;
            try { _reader?.Close(); } catch { }
            try { _writer?.Close(); } catch { }
            try { _sslStream?.Close(); } catch { }
            try { _socket?.Close(); } catch { }
        }
    }
}
